### PR TITLE
Pa3 update

### PR DESCRIPTION
**Pull request by tbsteinb**
This will update the assignment to have the students implement reference counting instead of concurrency.

Some potential improvements might be to have students run all the test cases between Parts I and II to compare both methods of GC (this may need some additional test cases that fail with reference counting but pass with copy collection, ones that cause fragmentation are probably easier to write).

The Extra Credit section still needs updated, I just wasn't quite sure what to put there or if it should just be removed (Mark+Sweep with a command line flag to use it instead of copy collection could be interesting).

### CODE REVIEW

File: pa/3.md [hunk 1]
```diff
@@ -25,11 +25,29 @@ The assignment has two parts. In the first, you will implement a copying garbage
...
+2. Execute the file, using reference counting for any data on the heap;
+3. Print the program's result to `stdout` using the textual encoding described in PA2 (`Vi32(-3)` for the integer `-3`, etc.). 
+
+More specific to reference counting, you will also need to meet the following guidelines:
+
+4. Write at least 2 unit tests specific to reference counting:
+    - One should verify that your reference counting implementation removes data with 0 references
+    - One should verify that the data you clear from the heap is used by Alloc if a spot exists with enough spaces
+5. When clearing spots from the heap, do not change the size of the heap, reference counting does not de-fragment memory
+6. If your VM cannot complete allocation, even with the space freed by the reference counter, then your VM should exit with a nonzero exit code
```
Comment (gstew5) [inline pa/3.md]: We should say something here about the fact that they don't need to collect cycles.

File: pa/3.md [hunk 2]
```diff
@@ -134,196 +152,7 @@ The program that generated `fact2.o` was:
...
-
-```
-> ./vm tests/conc-spawn3.o
-baVunit
-```
-
-depending on your scheduling policy (whether you run `f a` or `f b` first).
-
-## Extra Credit
+## Extra Credit (Should we add the old Part II as extra credit or just remove this?)
```
Comment (gstew5) [inline pa/3.md]: Let's just say no extra credit.

File: pa/3.md [hunk 3]
```diff
@@ -25,11 +25,29 @@ The assignment has two parts. In the first, you will implement a copying garbage
...
+1. Accept the name of a GrumpyVM bytecode `<filename.o>` as its first command-line argument;
+2. Execute the file, using reference counting for any data on the heap;
+3. Print the program's result to `stdout` using the textual encoding described in PA2 (`Vi32(-3)` for the integer `-3`, etc.). 
+
+More specific to reference counting, you will also need to meet the following guidelines:
+
+4. Write at least 2 unit tests specific to reference counting:
+    - One should verify that your reference counting implementation removes data with 0 references
+    - One should verify that the data you clear from the heap is used by Alloc if a spot exists with enough spaces
+5. When clearing spots from the heap, do not change the size of the heap, reference counting does not de-fragment memory
```
Comment (gstew5) [inline pa/3.md]: Period after "memory"

File: pa/3.md (hunk 1 above)
Comment (gstew5) [inline pa/3.md]: Period "." here too.

**Review by gstew5 — CHANGES_REQUESTED** [review]

File: pa/3.md (hunk 2 above)
Comment (tbsteinb) [reply pa/3.md]: This section was now removed.

File: pa/3.md (hunk 1 above)
Comment (tbsteinb) [reply pa/3.md]: Does the most recent addition work, or should it be a little more clear?

**Comment by tbsteinb** [comment]
While this is normally bad form, the students need to be able to start the assignment so I'm going to merge this PR. Changes can be added on an as-needed basis.

**Pull request merged**